### ISSUE TITLE
Suggestion: narrower parameters for NumberConstructor

### ISSUE THREAD

**Issue by danvk** [Microsoft/TypeScript#20754]
**TypeScript Version:**  2.6.2

**Code**

```ts
function foo() {
    return 1234;
}

const num = Number(foo);
// intent was probably:
// const num = foo();
```

**Expected behavior:** `Number(foo)` should be an error.

**Actual behavior:** Type checks. `num` is `NaN` at runtime.

I assume this is WAI, but figured I'd ask. This came up when I got a type error in a math expression, something like `2 * box.width`. I didn't read the error closely and assumed the issue was that `box.width` was a string. So I made it `2 * Number(box.width)`. This made the error go away, but introduced `NaN`s at runtime.

`Number` is declared as:

```ts
const Number: NumberConstructor(value?: any) => number
```

could we make this narrower? maybe

```ts
const Number: NumberConstructor(value?: string|number|Date|boolean) => number
```

**Comment by RyanCavanaugh** [comment]
The "uncalled function" problem is more broad than `Number(x)`; we speculate about a type that means "anything except a callable function" that would be the input type to a lot of coercion operators, though the existence of classes with static properties really muddies the waters. It's arguable that the `object` type should have excluded `function` types, even though `(function() { }) instanceof Object` is `true`.

Anyway "pass it to `Number`" is sort of the "big hammer" to coerce anything to a number and we'd rather not break legit (if odd) uses of that function.

In conclusion, JavaScript is a land of contrasts.
